### DISCUSSION TITLE
What's the status of open energy models in the US?

### DISCUSSION THREAD

**Discussion by zaneselvans** [catalyst-cooperative/pudl#2346] [Help Me!] [answered]
@bendnorman asked in Slack and it seemed worth copy-pasting here:
> What's the status of open energy models in the US? I saw [this RMI](https://rmi.org/economic-tides-just-turned-for-states/) report uses EI’s Energy Policy Simulator. There is also [Gen X from Princeton](https://energy.mit.edu/genx/). Any idea if the community is converging on one of these tools?

**Comment by zaneselvans** [comment, marked as answer]
The open energy system models that I’ve most frequently come across are
* [Switch 2.0](https://github.com/switch-model/switch) (Python, Apache 2.0, @switch-model)
* [Temoa](https://github.com/TemoaProject/oeo) (Python, no license in that repo, @TemoaProject)
* [GenX](https://github.com/GenXProject/GenX) (Julia, GPL, @JesseJenkins)
* [PyPSA](https://github.com/PyPSA/PyPSA) (Python, MIT, @nworbmot)
* [GridPath](https://github.com/blue-marble/gridpath) (Python, Apache 2.0, @anamileva) is the one GridLab has been working with / on.
* NREL’s ReEDS model isn’t really open. They’ll let you look at a private version of the source code, but the license prevents anybody from doing anything with it.

@grgmiller used Switch 2.0 in his work [on 24/7 power for that CCA in NorCal](https://www.volts.wtf/p/an-energy-provider-attempts-to-achieve). Temoa is being used by the Open Energy Outlook. @gschivley / @PowerGenome is (I think) currently attempting to support GenX, Switch, and Temoa, which means they could all be consuming some PUDL data.

My understanding is that all of these models end up doing “mixed integer linear programming” using optimization methods to solve a big set of coupled linear equations, and they typically end up being computationally limited in the complexity of the system they can represent.

They all depend on an external solver, and there are both open and proprietary versions, and unfortunately IIRC the open ones are like 10-100x slower than the proprietary ones. I think the big proprietary solvers are IBM’s CPLEX and Gurobi and the big open source solvers are [GLPK](https://github.com/firedrakeproject/glpk) and [lp_solve](https://lpsolve.sourceforge.net/5.5/) (whoa, SourceForge a blast from the past).

And there are sooo many more open modeling frameworks:
* [Here’s a journal article comparing 31 of them](https://www.sciencedirect.com/science/article/abs/pii/S1364032118307743)
* [Wikipedia page](https://en.wikipedia.org/wiki/Open_energy_system_models)
* [OpenMod Wiki](https://wiki.openmod-initiative.org/wiki/Open_Models)

Whether they’re usable locally or need massive resources will depend on how complex the system is you’re trying to model. Toy models run fine locally, but if you wanted to simulate a whole utility service territory or ISO or interconnect region, you have to do some combination of simplifying the system you’re modeling, and adding more compute resources, or getting access to a faster commercial solver. My guess is for real-world modeling, if you were doing a lot of it, it would quickly become worth finding a commercial solver license since you could cut the compute time down dramatically.  Last I checked the were still traditional “$40,000 per seat” licenses. But maybe you can pay by the hour or minute or second now via cloud services, like with accessing GPUs/TPUs? (Edit: It looks like [Google does provide some kind of API access to powerful solvers](https://developers.google.com/optimization))

Open energy system models definitely suffer from Not Invented Here — everybody wants to build their own model from scratch, so there’s 100 of them with limited functionality that are hard to maintain, instead of 5 that take different basic approaches and have a modular design with lots of maintainers. But hopefully it’ll get there eventually.  I think having open data available as inputs, and getting validated through standard benchmarking scenarios that allow direct comparisons between models would help put a few of them out in the front.

**Reply by grgmiller** [reply]
I'll note that the model used for the Peninsula Clean Energy analysis was built on top of the Switch 2.0 framework but is it's own distinct model called [MATCH](https://github.com/pencleanenergy/MATCH-model), tailored toward voluntary/PPA-based power procurement.

**Comment by nworbmot** [comment]
That's a pretty good summary. Perhaps a few additions:
- The folks @Breakthrough-Energy have also developed [PowerSimData](https://github.com/Breakthrough-Energy/PowerSimData) to wrap [REISE.jl](https://github.com/Breakthrough-Energy/REISE.jl) for US modelling, and the data for the US is [ported to PyPSA](https://github.com/PyPSA/pypsa-breakthroughenergy-usa). The energy modelling team @Breakthrough-Energy has now been wound down, so it would be great if the community developed it further.
- [HiGHS](https://github.com/ERGO-Code/HiGHS/) is an up-and-coming OS solver, see [performance for energy problems](https://forum.openmod.org/t/open-source-highs-solver-performance-boost-for-energy-system-models/2922). I think the new version is even better, we will create some benchmarks soon.
- There are several reviews of energy system frameworks, my favourite is [this one](https://doi.org/10.1016/j.rser.2018.11.020) "this work assesses 31 mostly open source tools based on 81 functions for their maturity. The result shows that available open source tools such as Switch, TEMOA, OSeMOSYS, and pyPSA are mature enough based on a function comparison with commercial or proprietary tools for serious use"
